### DISCUSSION TITLE
Help customizing Button via design token. v4

### DISCUSSION THREAD

**Discussion by jeremy-langevin-twc** [primefaces/community#2078] [PrimeVue]
Hello, I am able to use design tokens to easily adjust the colors of my navigation item application wide. However, when attempting to do something similar with buttons, I am not able to accomplish this. Could someone help steer me in the right direction?

`Version: ^4.0.0-beta.4`
Using styled mode, here is my theme:

```
import { palette } from 'primevue/themes';

const primary = palette('#0092C3');
const secondary = palette('#86B240');

export default {
    primitive: {
        secondary: secondary
    },
    semantic: {
        primary: primary,
        colorScheme: {
            light: {
                navigation: {
                    item: {
                        focusBackground: '{primary.50}',
                        color: '{primary.color}',
                        focusColor: '{primary.hover.color}'
                    }
                },
                button: {
                    color: '{secondary.500}',
                    primary: {
                        color: '{secondary.500}',
                        backgroundColor: '{secondary.100}'
                    },
                    item: {
                        background: '{secondary.500}',
                        color: '{slate.500}'
                    }
                }
            }
        }
    }
};
```

**Comment by bobby-driggs** [comment]
Try nesting under a `root` property, as shown in their Lara theme? 

https://github.com/primefaces/primevue/blob/e275149c3d2f74d6b8fcaff057f30023178a3102/components/lib/themes/lara/button/index.js#L35

**Comment by svwezel** [comment]
I also can't get it to work. (Also tried nesting under a root component)

**Comment by nickwall** [comment]
Hi.

I need to theme all elements to the corporate pallet.  I've changed the default primary pallet, default hovers, focus, etc. all good.

But changing components, I'm struggling a bit with getting this to work.

I have followed the example as described in video here around the 10:15 mark:
https://www.youtube.com/watch?v=J3KFw5sih98

..whatever I do, I can not get it to work for me.

Below is an example of what I have done in my preset file to test various things.

  components: {
    button: {
      borderRadius: '2.5rem', // works
      labelFontWeight: 'bold', // works
      paddingX: '3.5rem', // not work
      paddingy: '1.5rem', // not work
      primaryHoverColor: '{primary.100}', // not work
      primaryColor: '#26b92e', // not work
      primaryBackground: '#1e54c7', // not work
      primaryBorderColor: '{primary.950}' // not work
    }

At this stage, I just want to change the background colour of the primary button and change it's border so i understand how it all fits together.

If I inspect the element in dev tools, I can see that the config file sets the ":root" styles somehow and I can see the props I'm setting, e.g. primaryBackground: '#1e54c7', // not work  getting transformed to --p-button-primary-background: #1e54c7'   but it's getting disabled by a property superseding it.
--p-button-primary-background: var(--p-primary-color);

Has anyone got any idea how to get this to work to any avenues to explore?  

Thanks in advance.

**Reply by j0rgedev** [reply]
Hey @nickwall could you please provide a reproduction or share your custom preset?

**Reply by nickwall** [reply]
Hi,

Thanks for getting back to me.

Here is the preset I am experimenting with.  as mentioned, I can't get back ground color, border, etc. to change.

{
  semantic: {
    primary: {
      50: '#f9e6f0',
      100: '#f2cde1',
      200: '#e6a3c7',
      300: '#d97aaf',
      400: '#cc5196',
      500: '#bf287d',
      600: '#a31f66',
      700: '#86164f',
      800: '#b02775',
      900: '#590a2b',
      950: '#2d0515'
    }
  },
  components: {
    button: {
      borderRadius: '2.5rem', // works
      labelFontWeight: 'bold', // works
      paddingX: '3.5rem', // works
      paddingY: '1.5rem', // works
      primaryHoverColor: '{primary.100}', // not work
      primaryColor: '#26b92e', // not work
      primaryBackground: '#1e54c7', // not work
      primaryBorderColor: '{primary.950}' // not work
    }
  }
}

**Reply by j0rgedev** [reply]
@nickwall I noticed that your preset isn't working as expected because it lacks the correct button component configuration. This issue is understandable as it's not yet well documented, but here's a way to handle it. All color-related settings for components should be structured in the following scheme:
```javascript
   ...
   colorScheme: {
        light: {
            root: {
                primary: {
                    // set custom properties here. These can change depending on the component. (I'll later explain it)
                    ...
                },
            },
        },
        dark:
            root: {
                primary: {
                    // same as above
                    ...
                },
            },
    }
```

## How to Define Component Properties
You might be wondering how to figure out which properties to set for each component. There are two main approaches:

- **Theming Tab**: Each component has a theming tab where you can see the design tokens that you need to set for your custom design. From my point of view, it's too generic, so I prefer the next approach.

- **Source Code**: Since the documentation is still lacking, the best approach is to look directly at the source code. You can find all the component schemas at this path: [PrimeVue Component Schemas](https://github.com/primefaces/primevue/blob/master/packages/themes/src/presets/aura/). This will give you a clear reference to set your custom design without errors.

To make it clearer, here's an example of how your custom preset should look when it's correctly structured:
```javascript
import { definePreset } from '@primevue/themes';
import Aura from '@primevue/themes/aura';

const CustomPreset = definePreset(Aura, {
  semantic: {
    primary: {
      50: '#f9e6f0',
      100: '#f2cde1',
      200: '#e6a3c7',
      300: '#d97aaf',
      400: '#cc5196',
      500: '#bf287d',
      600: '#a31f66',
      700: '#86164f',
      800: '#b02775',
      900: '#590a2b',
      950: '#2d0515',
    },
  },
  components: {
    button: {
      borderRadius: '2.5rem',
      labelFontWeight: 'bold',
      paddingX: '3.5rem',
      paddingY: '1.5rem',
      colorScheme: {
        light: {
          root: {
            primary: {
              color: '#26b92e',
              hoverColor: '{primary.100}',
              background: '#1e54c7',
              borderColor: '{primary.950}',
            },
          },
        },
      },
    },
  },
});

export default CustomPreset;
```
This setup ensures that all your button components and color schemes are correctly placed in the right structure, which should make your preset function as expected. I hope I explained everything clearly. Let me know if you have any questions. 😁

**Reply by nickwall** [reply]
Great explanation.   Got it working now.  Have a great rest of your weekend.

**Reply by llamaza** [reply]
Server must be restarted or npm run build to see customization changes.

**Reply by jkocyigi** [reply]
Thank you so much! spent half a day trying to this out figure out 👍

**Reply by hydesight** [reply]
Thanks alot. You explained it the best.

**Reply by j0rgedev** [reply]
# Update on PrimeVue Themes  

Hey there! Glad my comment helped, but a lot has changed since then.  

- **Themes** has now become part of the [primeuix repository](https://github.com/primefaces/primeuix), which contains shared utilities used across all PrimeFaces libraries (PrimeNG, PrimeReact, etc.). This includes functionalities for forms, locale and more. Because of that, since **PrimeVue 4.3.0**, importing your theme from `@primevue/themes/xxxxx` is now deprecated.  
- **Version 4.3.0** also brought better TypeScript support, so forget what I said about checking the source code. It’s not needed anymore!  

## Updated Custom Preset  
Here's how your custom preset should look now:  

### `app-theme.ts` _(Make sure it’s a TS file)_  
```ts
import { definePreset } from '@primeuix/themes';
import Aura from '@primeuix/themes/aura';

// Everything else stays the same...
```

With this small change, you'll have better completion for each component, so now you won't miss any design token configurations! Hope this helps 😁

**Reply by dnszero** [reply]
Thank you for this! I _finally_ get it now. 

PrimeFaces needs to link your comment in their docs!

**Comment by dutchcube** [comment]
@j0rgedev Thanks a lot for your answer! Understanding it a lot more now.

**Comment by JoeyRoeters** [comment]
The design tokens are now included in the new primeuix/themes library. With TS make the component variable typed like: 
`const componentTokens: ComponentsDesignTokens = {}`

with `import { ComponentsDesignTokens } from '@primeuix/themes/types'`

Then you will get autocompletion and it will be much easier to find them
